File: dataset_reduzido/avatar/Python/Code/atcoder_AGC005_B.py
root = lambda x : x if par [ x ] == x else root ( par [ x ] )
same = lambda x , y : root ( x ) == root ( y )
def unite ( * z ) :
    x , y = map ( root , z )
    if x != y :
        if rank [ x ] < rank [ y ] :
            par [ x ] = y
            size [ y ] += size [ x ]
        else :
            par [ y ] = x
            size [ x ] += size [ y ]
            if rank [ x ] == rank [ y ] : rank [ x ] += 1
n , * a = map ( int , open ( 0 ).read ( ).split ( ) )
* par , = range ( n )
rank = [ 1 ] * n
size = [ 1 ] * n
b = [ 0 ] * n
for i , t in enumerate ( a ) : b [ t - 1 ] = i
c = 0
for k in b [ : : - 1 ] :
    l = r = 1
    if k > 0 and a [ k - 1 ] > a [ k ] :
        l += size [ root ( k - 1 ) ]
        unite ( k - 1 , k )
    if k + 1 < n and a [ k + 1 ] > a [ k ] :
        r += size [ root ( k + 1 ) ]
        unite ( k + 1 , k )
    c += l * r * a [ k ]
print ( c )

File: output/StarCoder/avatar/Python/Java/atcoder_AGC005_B.java
```java
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String[] input = scanner.nextLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int[] a = Arrays.stream(input).skip(1).mapToInt(Integer::parseInt).toArray();
        
        int[] par = new int[n];
        for (int i = 0; i < n; i++) {
            par[i] = i;
        }
        
        int[] rank = new int[n];
        Arrays.fill(rank, 1);
        
        int[] size = new int[n];
        Arrays.fill(size, 1);
        
        int[] b = new int[n];
        for (int i = 0; i < n; i++) {
            b[a[i] - 1] = i;
        }
        
        long c = 0;
        for (int k = n - 1; k >= 0; k--) {
            int l = 1, r = 1;
            if (k > 0 && a[k - 1] > a[k]) {
                l += size[root(par, k - 1)];
                unite(par, rank, size, k - 1, k);
            }
            if (k + 1 < n && a[k + 1] > a[k]) {
                r += size[root(par, k + 1)];
                unite(par, rank, size, k + 1, k);
            }
            c += (long) l * r * a[k];
        }
        
        System.out.print(c);
    }

    private static int root(int[] par, int x) {
        return par[x] == x ? x : root(par, par[x]);
    }

    private static boolean same(int[] par, int x, int y) {
        return root(par, x) == root(par, y);
    }

    private static void unite(int[] par, int[] rank, int[] size, int x, int y) {
        x = root(par, x);
        y = root(par, y);
        if (x != y) {
            if (rank[x] < rank[y]) {
                par[x] = y;
                size[y] += size[x];
            } else {
                par[y] = x;
                size[x] += size[y];
                if (rank[x] ==